Bug 1875909 - Suggest: Add support for mobile AMP suggestions

tiftran opened the pull request:
### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

tiftran commented on components/suggest/src/db.rs:
```diff
@@ -186,7 +186,9 @@ impl<'a> SuggestDao<'a> {
             .try_fold(vec![], |mut acc, provider| {
                 let suggestions = match provider {
-                    SuggestionProvider::Amp => self.fetch_amp_suggestions(query),
+                    SuggestionProvider::Amp | SuggestionProvider::AmpMobile => {
+                        self.fetch_amp_suggestions(query, **provider)
+                    }
```
So I tried my best to recycle the existing bits as much as I can, but as a result, this look a little goofy since at the point of `fetch_amp_suggestions` we don't know which provider it came from.

tiftran commented on components/suggest/src/db.rs:
```diff
@@ -731,11 +754,16 @@ impl<'a> SuggestDao<'a> {
             self.scope.err_if_interrupted()?;
             let common_details = suggestion.common_details();
-            let provider = suggestion.provider();
+            let provider = if let SuggestRecord::AmpMobile = fields {
+                SuggestionProvider::AmpMobile
+            } else {
+                suggestion.provider()
+            };
```
the clunky-ness continues here, since  `suggestion.providers()` differentiates amp and wikipedia by the advertiser (I think), and AMP mobile and AMP are the exact same in terms of structure?

linabutler replied:
Yeah, I think that's why I'm leaning toward having a separate `insert_amp_mobile_suggestions` method. `insert_amp_wikipedia_suggestions` has to disentangle AMP Desktop and Wikipedia suggestions, and with the addition of AMP Mobile, this method now handles three different cases. A separate method would bring it down to two.

ncloudioj commented on components/suggest/src/db.rs:
```diff
@@ -731,11 +754,16 @@ impl<'a> SuggestDao<'a> {
         &mut self,
         record_id: &SuggestRecordId,
         suggestions: &[DownloadedAmpWikipediaSuggestion],
+        fields: SuggestRecord,
```
Nit: since this is only used to determine if the record is of type `amp_mobile_suggestions`, we can just pass in a boolean `is_amp_mobile`.

```suggestion
        is_amp_mobile: bool,
```

```rs
dao.insert_amp_wikipedia_suggestions(
   record_id,
   suggestions,
   matches!(fields, SuggestRecrod::AmpMobile)
);
```

linabutler replied:
I might actually suggest having a separate `dao.insert_amp_mobile_suggestions` method, at the risk of duplicating a bit of code!

The reason I'd suggest that is that AMP Desktop and Wikipedia suggestions are commingled in the same attachment, _but_ AMP Mobile only has mobile suggestions! So maybe we can have a method on `SuggestDao` like:

```rs
fn insert_amp_mobile_suggestions(&mut self, record_id: &SuggestRecordId, suggestions: &[DownloadedAmpSuggestion]) -> Result<()>
```

...That doesn't have to worry about Wikipedia.

ncloudioj replied:
Ticket for the split: https://bugzilla.mozilla.org/show_bug.cgi?id=1879962

ncloudioj commented on components/suggest/src/suggestion.rs:
```diff
@@ -33,6 +33,20 @@ pub enum Suggestion {
         raw_click_url: String,
         score: f64,
     },
+    AmpMobile {
```
Since AMP suggestions share the same structure on desktop and mobile, I think we can use `Amp` for both. Let's add a doc string to to the `Amp` variant to note this.

linabutler replied:
Hmmm, I'm not sure about this one! Right now, there's a 1:1 correspondence between the variants of `SuggestionProvider` and `Suggestion`—so if you ask for `SuggestionProvider::Amp` and `SuggestionProvider::Wikipedia`, you'll get `Suggestion::Amp` and `Suggestion::Wikipedia` back.

Even though the AMP mobile suggestions have the same structure, I think passing `SuggestionProvider::AmpMobile` and `SuggestionProvider::Wikipedia`, and getting `Suggestion::Amp` and `Suggestion::Wikipedia` back, feels a bit too special case-y for me 🤔

linabutler replied:
I thought a bit more about this one, and now I like @ncloudioj's suggestion more :joy:

There's no reason why a `SuggestionProvider` couldn't return different types of `Suggestion`s, and I could see us doing this in the future.

ncloudioj commented on components/suggest/src/db.rs:
```diff
@@ -276,20 +282,37 @@ impl<'a> SuggestDao<'a> {
-                            score,
-                        })
+                        if provider == SuggestionProvider::Amp {
+                            Ok(Suggestion::Amp {
```
As noted below, we can use `Suggestion::Amp` for both mobile and desktop.

ncloudioj commented on components/suggest/src/store.rs:
```diff
@@ -3558,6 +3558,191 @@ mod tests {
+                "keyword = `las `; amp and amp mobile",
+                SuggestionQuery {
+                    keyword: "las".into(),
+                    providers: vec![SuggestionProvider::Amp, SuggestionProvider::AmpMobile],
```
Note that, in practice, consumers would only issue queries against either of those two providers, and it's up to the consumers to enforce that. But I guess having a test here to query both is fine for completeness.

ncloudioj reviewed (changes requested):
Looks good, just a few comments for simplification.

linabutler reviewed (changes requested):
Almost there! My suggestions are a little different than @ncloudioj's, but I'm happy to discuss them more, and open to having my mind changed! ☺️

* I think we can keep `Suggestion::AmpMobile` and `Suggestion::Amp` separate.
* I'd suggest adding a new `AmpSuggestionType` enum to `db.rs`, and using it in `fetch_amp_suggestions`.
* I'd also suggest adding a separate `SuggestDao::insert_amp_mobile_suggestions` method, because AMP Mobile doesn't need to untangle Wikipedia suggestions.

ncloudioj commented on components/suggest/src/db.rs:
```diff
@@ -207,7 +209,11 @@ impl<'a> SuggestDao<'a> {
+    pub fn fetch_amp_suggestions(
+        &self,
+        query: &SuggestionQuery,
+        provider: SuggestionProvider,
```
YES! +1. Let's [make-invalid-states-unrepresentable](https://geeklaunch.io/blog/make-invalid-states-unrepresentable/)!

Actually, I was gonna suggest adding this exact enum `AmpSuggestionType` to the `Provider` so that consumers have to pick either variant (but not both!) when they query. Then I realized that they could still "violate" that as they could pass both `Mobile` and `Desktop` to `SuggestionQuery::providers` 😂.

linabutler replied:
Oooh, thank you for that link, it's going into my bookmarks for sure! 🏆 I think [this Aphyr article](https://aphyr.com/posts/341-hexing-the-technical-interview) (it's a _read_! :joy:) was the first time I heard that expression:

> “But… those aren’t even the same type. That’s… that’s illegal.”
>
> “If it were meant to be illegal,” you remind him sagely, “Sun Microsystems would have made it unrepresentable.”

I liked your second point, too:

> Then I realized that they could still "violate" that as they could pass both `Mobile` and `Desktop` to `SuggestionQuery::providers`.

Yeah, it's tricky; I think this is the first time that we're segmenting suggestions by platform like this! I'd be curious to see if we end up with similar cases in the future—and if a more fluent way of expressing this separation emerges from those other cases! For now, it feels like the list of providers is the most sensible abstraction we have, but that might change as we add providers and new types of filters.

tiftran commented:
okie doke, i think we ended up converging in agreement with changes, if i read everything correctly.

- added an enum, for desktop and mobile
- removed `AmpMobile` as a suggestion type
- added `SuggestDao::insert_amp_mobile_suggestions`

linabutler commented on components/suggest/src/suggestion.rs:
```diff
@@ -16,6 +16,11 @@ const TIMESTAMP_TEMPLATE: &str = "%YYYYMMDDHH%";
 const TIMESTAMP_LENGTH: usize = 10;
 
+/// Suggestion Types for Amp
+pub enum AmpSuggestionType {
```
```suggestion
pub(crate) enum AmpSuggestionType {
```

Nit: Let's stick to `pub` for things that we expose in the UDL, and `pub(crate)` for things that we use within the crate.

linabutler commented on components/suggest/src/store.rs:
```diff
@@ -3999,6 +4004,191 @@ mod tests {
+                  "keywords": ["la", "las", "lasa", "lasagna", "lasagna come out tomorrow"],
+                  "title": "Desktop - Lasagna Come Out Tomorrow",
+                  "url": "https://www.lasagna.restaurant",
+                  "icon": "2",
```
```suggestion
                  "icon": "3",
```

linabutler commented on components/suggest/src/store.rs:
```diff
@@ -3999,6 +4004,191 @@ mod tests {
+                   "keywords": ["la", "las", "lasa", "lasagna", "lasagna come out tomorrow"],
+                   "title": "Mobile - Lasagna Come Out Tomorrow",
+                   "url": "https://www.lasagna.restaurant",
+                   "icon": "2",
```
```suggestion
                   "icon": "3",
```

Just to make sure that we ingest icons correctly for mobile, too ☺️

linabutler reviewed (approved):
Looks great, thanks! :shipit:

ncloudioj reviewed (approved):
LGTM, thanks!

Once AMP mobile suggestions are ready in production (pending on AMP's confirmation), we will need two tiny client changes to swap in the new AMP mobile provider.
